[mlir][gpu][bufferization] Implement BufferDeallocationOpInterface for gpu.terminator

maerhart opened the pull request:
This is necessary to support deallocation of IR with gpu.launch operations because it does not implement the RegionBranchOpInterface. Implementing the interface would require it to support regions with unstructured control flow and produced arguments/results.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferDeallocationOpInterface.h:
```diff
@@ -205,6 +205,19 @@ class DeallocationState {
+namespace deallocation_impl {
+/// Insert a `bufferization.dealloc` operation right before 'op' which has to be
+/// a terminator without any successors. Note that it is not required to have
+/// the ReturnLike trait attached. The MemRef values in the 'operands' argument
```
nit: you can use backticks for variable/function names

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferDeallocationOpInterface.h:
```diff
@@ -205,6 +205,19 @@ class DeallocationState {
+/// the ReturnLike trait attached. The MemRef values in the 'operands' argument
+/// will be added to the list of retained values and their updated ownership
+/// values will be appended to the 'updatedOperandOwnerships' list. 'op' is not
+/// modified in any way. Returns failure or the original 'op'.
```
Mention what the meaning of "failure" is and when it would be returned.

maerhart committed:
[mlir][gpu][bufferization] Implement BufferDeallocationOpInterface for gpu.terminator

This is necessary to support deallocation of IR with gpu.launch operations
because it does not implement the RegionBranchOpInterface. Implementing the
interface would require it to support regions with unstructured control flow
and produced arguments/results.
